Configuration des chemins

In [17]:
import os, sys; sys.path.append(os.path.abspath(os.path.join(os.getcwd(), "..")))


Imports des modules

In [49]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from src.data_loader import load_recipes, load_interactions, sanity_check

**Test du chargement des données**

Sanity check

In [19]:
sanity_check()

ROOT: C:\Users\ABDILLAHIOMARDJAMA\Downloads\EDA_Project
RAW : C:\Users\ABDILLAHIOMARDJAMA\Downloads\EDA_Project\data\raw exists: True
CONTENU RAW:
  • .gitkeep
  • RAW_interactions.csv
  • RAW_recipes.csv


Chargement des datasets

In [102]:
recipes = load_recipes()
interactions = load_interactions()

✅ data\raw\RAW_recipes.csv  shape=(231637, 12)
✅ data\raw\RAW_interactions.csv  shape=(1132367, 5)


In [113]:
recipes['contributor_id'].value_counts().head()

contributor_id
89831    3118
37779    2553
37449    2493
1533     1595
58104    1522
Name: count, dtype: int64

### EXPLORATION INITIALE DES DONNÉES  

Cette section explore **systématiquement** nos datasets pour comprendre leur structure, qualité et contenu.

### 3.1 Vue d'Ensemble des Datasets

In [35]:
#  DIMENSIONS ET FORME DES DATASETS
print(" RÉSUMÉ DES DATASETS")
print("=" * 50)
print(f" Recettes:")
print(f"   • Lignes: {recipes.shape[0]:,}")
print(f"   • Colonnes: {recipes.shape[1]}")

print(f"\n Interactions:")
print(f"   • Lignes: {interactions.shape[0]:,}")
print(f"   • Colonnes: {interactions.shape[1]}")


 RÉSUMÉ DES DATASETS
 Recettes:
   • Lignes: 231,637
   • Colonnes: 12

 Interactions:
   • Lignes: 1,132,367
   • Colonnes: 5


In [51]:
#  STRUCTURE DES DATASETS
print(" STRUCTURE DES RECETTES")
print("=" * 40)
print(f"Colonnes: {list(recipes.columns)}")
print("\nTypes de données:")
display(recipes.dtypes.to_frame('Type'))

print("\n STRUCTURE DES INTERACTIONS") 
print("=" * 40)
print(f"Colonnes: {list(interactions.columns)}")
print("\nTypes de données:")
display(interactions.dtypes.to_frame('Type'))

 STRUCTURE DES RECETTES
Colonnes: ['name', 'id', 'minutes', 'contributor_id', 'submitted', 'tags', 'nutrition', 'n_steps', 'steps', 'description', 'ingredients', 'n_ingredients']

Types de données:


Type
name            object
id               int64
minutes          int64
contributor_id   int64
submitted       object
tags            object
nutrition       object
n_steps          int64
steps           object
description     object
ingredients     object
n_ingredients    int64


 STRUCTURE DES INTERACTIONS
Colonnes: ['user_id', 'recipe_id', 'date', 'rating', 'review']

Types de données:


Type
user_id     int64
recipe_id   int64
date       object
rating      int64
review     object

Description des données

In [31]:
#  APERÇU DES PREMIÈRES LIGNES
print(" APERÇU DES RECETTES (5 premières lignes)")
print("=" * 50)
display(recipes.head())

print("\n APERÇU DES INTERACTIONS (5 premières lignes)")
print("=" * 50)
display(interactions.head())

 APERÇU DES RECETTES (5 premières lignes)


name      id  minutes  \
0  arriba   baked winter squash mexican style  137739       55   
1            a bit different  breakfast pizza   31490       30   
2                   all in the kitchen  chili  112140      130   
3                          alouette  potatoes   59389       45   
4          amish  tomato ketchup  for canning   44061      190   

   contributor_id   submitted  \
0           47892  2005-09-16   
1           26278  2002-06-17   
2          196586  2005-02-25   
3           68585  2003-04-14   
4           41706  2002-10-25   

                                                tags  \
0  ['60-minutes-or-less', 'time-to-make', 'course...   
1  ['30-minutes-or-less', 'time-to-make', 'course...   
2  ['time-to-make', 'course', 'preparation', 'mai...   
3  ['60-minutes-or-less', 'time-to-make', 'course...   
4  ['weeknight', 'time-to-make', 'course', 'main-...   

                                    nutrition  n_steps  \
0       [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1   [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3   [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4   [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   

                                               steps  \
0  ['make a choice and proceed with recipe', 'dep...   
1  ['preheat oven to 425 degrees f', 'press dough...   
2  ['brown ground beef in large pot', 'add choppe...   
3  ['place potatoes in a large pot of lightly sal...   
4  ['mix all ingredients& boil for 2 1 / 2 hours ...   

                                         description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  this modified version of 'mom's' chili was a h...   
3  this is a super easy, great tasting, make ahea...   
4  my dh's amish mother raised him on this recipe...   

                                         ingredients  n_ingredients  
0  ['winter squash', 'mexican seasoning', 'mixed ...              7  
1  ['prepared pizza crust', 'sausage patty', 'egg...              6  
2  ['ground beef', 'yellow onions', 'diced tomato...             13  
3  ['spreadable cheese with garlic and herbs', 'n...             11  
4  ['tomato juice', 'apple cider vinegar', 'sugar...              8


 APERÇU DES INTERACTIONS (5 premières lignes)


user_id  recipe_id        date  rating  \
0    38094      40893  2003-02-17       4   
1  1293707      40893  2011-12-21       5   
2     8937      44394  2002-12-01       4   
3   126440      85009  2010-02-27       5   
4    57222      85009  2011-10-01       5   

                                              review  
0  Great with a salad. Cooked on top of stove for...  
1  So simple, so delicious! Great for chilly fall...  
2  This worked very well and is EASY.  I used not...  
3  I made the Mexican topping and took it to bunk...  
4  Made the cheddar bacon topping, adding a sprin...

We have a total of 231637 recipes and each recipe is described by 12 variables. 
- id entier Identifiant unique de la recette
- name texte Nom de la recette
- minutes numérique (int) Temps total estimé pour la recette (préparation + cuisson)
- contributor_id entier Identifiant de l’utilisateur ayant publié la recette
- submitted date Date de soumission de la recette
- tags liste (chaîne de type "['tag1','tag2']")
- nutrition liste (chaîne de type "[val1, val2, ...]") Données nutritionnelles (généralement : calories, gras, sucre, sodium, protéines, etc.)
- n_steps entier Nombre d’étapes dans la recette
- steps liste (chaîne list-like) Description détaillée des étapes
- description texte Description libre de la recette
- ingredients liste (chaîne list-like) Liste des ingrédients
- n_ingredients entier Nombre d’ingrédients utilisés


#### 3.2 Qualité des Données

In [32]:
#  ANALYSE DES VALEURS MANQUANTES
print("VALEURS MANQUANTES DANS LES RECETTES")
print("=" * 45)
missing_recipes = recipes.isnull().sum()
total_recipes = len(recipes)

for col in missing_recipes.index:
    missing_count = missing_recipes[col]
    missing_pct = (missing_count / total_recipes) * 100
    if missing_count > 0:
        print(f"   • {col}: {missing_count:,} ({missing_pct:.1f}%)")
    else:
        print(f"    {col}: Aucune valeur manquante")

print(f"\n VALEURS MANQUANTES DANS LES INTERACTIONS")
print("=" * 48)
missing_interactions = interactions.isnull().sum()
total_interactions = len(interactions)

for col in missing_interactions.index:
    missing_count = missing_interactions[col]
    missing_pct = (missing_count / total_interactions) * 100
    if missing_count > 0:
        print(f"   • {col}: {missing_count:,} ({missing_pct:.1f}%)")
    else:
        print(f"    {col}: Aucune valeur manquante")

VALEURS MANQUANTES DANS LES RECETTES
   • name: 1 (0.0%)
    id: Aucune valeur manquante
    minutes: Aucune valeur manquante
    contributor_id: Aucune valeur manquante
    submitted: Aucune valeur manquante
    tags: Aucune valeur manquante
    nutrition: Aucune valeur manquante
    n_steps: Aucune valeur manquante
    steps: Aucune valeur manquante
   • description: 4,979 (2.1%)
    ingredients: Aucune valeur manquante
    n_ingredients: Aucune valeur manquante

 VALEURS MANQUANTES DANS LES INTERACTIONS
    user_id: Aucune valeur manquante
    recipe_id: Aucune valeur manquante
    date: Aucune valeur manquante
    rating: Aucune valeur manquante
   • review: 169 (0.0%)


In [34]:
#  DÉTECTION DES DOUBLONS
print(" ANALYSE DES DOUBLONS")
print("=" * 30)

# Doublons dans les recettes
recipes_duplicates = recipes.duplicated().sum()
print(f" Recettes dupliquées: {recipes_duplicates:,}")

if recipes_duplicates > 0:
    # Analyser les doublons par colonne
    recipes_dup_by_col = recipes.duplicated(subset=['name']).sum()
    print(f"   • Doublons par nom: {recipes_dup_by_col:,}")

# Doublons dans les interactions
interactions_duplicates = interactions.duplicated().sum()
print(f"\n Interactions dupliquées: {interactions_duplicates:,}")

if interactions_duplicates > 0:
    # Analyser les doublons par combinaison
    interactions_dup_user_recipe = interactions.duplicated(subset=['user_id', 'recipe_id']).sum()
    print(f"   • Doublons user_id + recipe_id: {interactions_dup_user_recipe:,}")

 ANALYSE DES DOUBLONS
 Recettes dupliquées: 0

 Interactions dupliquées: 0


Conversion

In [76]:
from src.preprocessing import convert_recipes_for_univariate, RecipeConversionConfig

recipes_converted = convert_recipes_for_univariate(recipes)

In [53]:
recipes_converted.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231637 entries, 0 to 231636
Data columns (total 21 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   name            231636 non-null  object        
 1   id              231637 non-null  int64         
 2   minutes         231637 non-null  int64         
 3   contributor_id  231637 non-null  category      
 4   submitted       231637 non-null  datetime64[ns]
 5   tags            231637 non-null  object        
 6   nutrition       231637 non-null  object        
 7   n_steps         231637 non-null  int64         
 8   steps           231637 non-null  object        
 9   description     226658 non-null  object        
 10  ingredients     231637 non-null  object        
 11  n_ingredients   231637 non-null  int64         
 12  calories        231637 non-null  float64       
 13  total_fat       231637 non-null  float64       
 14  sugar           231637 non-null  flo

In [71]:
recipes_converted.loc[:,['description', 'ingredients', 'steps', 'tags','minutes', 'n_ingredients', 'n_steps']].head(10)

description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  this modified version of 'mom's' chili was a h...   
3  this is a super easy, great tasting, make ahea...   
4  my dh's amish mother raised him on this recipe...   
5                                                NaN   
6  my italian mil was thoroughly impressed by my ...   
7  this recipe is posted by request and was origi...   
8                                                NaN   
9                                from ann hodgman's    

                                         ingredients  \
0  [winter squash, mexican seasoning, mixed spice...   
1  [prepared pizza crust, sausage patty, eggs, mi...   
2  [ground beef, yellow onions, diced tomatoes, t...   
3  [spreadable cheese with garlic and herbs, new ...   
4  [tomato juice, apple cider vinegar, sugar, sal...   
5  [milk, vanilla ice cream, frozen apple juice c...   
6  [fennel seeds, green olives, ripe olives, garl...   
7  [pork spareribs, soy sauce, fresh garlic, fres...   
8  [chocolate sandwich style cookies, chocolate s...   
9  [sugar, unsalted butter, bananas, eggs, fresh ...   

                                               steps  \
0  [make a choice and proceed with recipe, depend...   
1  [preheat oven to 425 degrees f, press dough in...   
2  [brown ground beef in large pot, add chopped o...   
3  [place potatoes in a large pot of lightly salt...   
4  [mix all ingredients& boil for 2 1 / 2 hours ,...   
5  [combine ingredients in blender, cover and ble...   
6  [toast the fennel seeds and lightly crush them...   
7  [in a medium saucepan combine all the ingredie...   
8  [crumble cookies into a 9-inch pie plate , or ...   
9  [preheat oven to 350 degrees, butter two 9x5" ...   

                                                tags  minutes  n_ingredients  \
0  [60-minutes-or-less, time-to-make, course, mai...       55              7   
1  [30-minutes-or-less, time-to-make, course, mai...       30              6   
2  [time-to-make, course, preparation, main-dish,...      130             13   
3  [60-minutes-or-less, time-to-make, course, mai...       45             11   
4  [weeknight, time-to-make, course, main-ingredi...      190              8   
5  [15-minutes-or-less, time-to-make, course, mai...        0              4   
6  [15-minutes-or-less, time-to-make, course, mai...       15              9   
7  [weeknight, time-to-make, course, main-ingredi...      120             22   
8  [weeknight, time-to-make, course, main-ingredi...      180              6   
9  [weeknight, time-to-make, course, main-ingredi...       70              9   

   n_steps  
0       11  
1        9  
2        6  
3       11  
4        5  
5        4  
6        4  
7       10  
8        8  
9       12

Valeur manquante

In [79]:
import importlib
from src import preprocessing
importlib.reload(preprocessing)

from src.preprocessing import generate_auto_description



In [92]:
type(recipes_converted["tags"].iloc[0])


list

In [81]:
recipes_converted["description_filled"] = recipes_converted.apply(
    generate_auto_description, axis=1
)
recipes_converted[["description", "description_filled"]].head(10)


description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  this modified version of 'mom's' chili was a h...   
3  this is a super easy, great tasting, make ahea...   
4  my dh's amish mother raised him on this recipe...   
5                                                NaN   
6  my italian mil was thoroughly impressed by my ...   
7  this recipe is posted by request and was origi...   
8                                                NaN   
9                                from ann hodgman's    

                                  description_filled  
0  autumn is my favorite time of year to cook! th...  
1  this recipe calls for the crust to be prebaked...  
2  this modified version of 'mom's' chili was a h...  
3  this is a super easy, great tasting, make ahea...  
4  my dh's amish mother raised him on this recipe...  
5  This is a 15 minutes or less recipe that is qu...  
6  my italian mil was thoroughly impressed by my ...  
7  this recipe is posted by request and was origi...  
8  This is a weeknight recipe that takes longer t...  
9                                from ann hodgman's

In [94]:
# --- création d'une nouvelle colonne longueur de description ---
from src.feature_engineering import add_recipe_features

recipes_enriched = add_recipe_features(recipes_converted)
recipes_enriched.head()


name      id  minutes contributor_id  \
0  arriba   baked winter squash mexican style  137739       55          47892   
1            a bit different  breakfast pizza   31490       30          26278   
2                   all in the kitchen  chili  112140      130         196586   
3                          alouette  potatoes   59389       45          68585   
4          amish  tomato ketchup  for canning   44061      190          41706   

   submitted                                               tags  \
0 2005-09-16  [60-minutes-or-less, time-to-make, course, mai...   
1 2002-06-17  [30-minutes-or-less, time-to-make, course, mai...   
2 2005-02-25  [time-to-make, course, preparation, main-dish,...   
3 2003-04-14  [60-minutes-or-less, time-to-make, course, mai...   
4 2002-10-25  [weeknight, time-to-make, course, main-ingredi...   

                                    nutrition  n_steps  \
0       [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1   [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3   [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4   [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   

                                               steps  \
0  [make a choice and proceed with recipe, depend...   
1  [preheat oven to 425 degrees f, press dough in...   
2  [brown ground beef in large pot, add chopped o...   
3  [place potatoes in a large pot of lightly salt...   
4  [mix all ingredients& boil for 2 1 / 2 hours ,...   

                                         description  ... total_fat  sugar  \
0  autumn is my favorite time of year to cook! th...  ...       0.0   13.0   
1  this recipe calls for the crust to be prebaked...  ...      18.0    0.0   
2  this modified version of 'mom's' chili was a h...  ...      22.0   32.0   
3  this is a super easy, great tasting, make ahea...  ...      17.0   10.0   
4  my dh's amish mother raised him on this recipe...  ...       1.0  337.0   

   sodium  protein  saturated_fat  carbohydrates  year  month  \
0     0.0      2.0            0.0            4.0  2005      9   
1    17.0     22.0           35.0            1.0  2002      6   
2    48.0     39.0           27.0            5.0  2005      2   
3     2.0     14.0            8.0           20.0  2003      4   
4    23.0      3.0            0.0           28.0  2002     10   

                                  description_filled  description_length  
0  autumn is my favorite time of year to cook! th...                  32  
1  this recipe calls for the crust to be prebaked...                  33  
2  this modified version of 'mom's' chili was a h...                  60  
3  this is a super easy, great tasting, make ahea...                  43  
4  my dh's amish mother raised him on this recipe...                  35  

[5 rows x 23 columns]

In [91]:
recipes_enriched.dtypes

name                          object
id                             int64
minutes                        int64
contributor_id              category
submitted             datetime64[ns]
tags                          object
nutrition                     object
n_steps                        int64
steps                         object
description                   object
ingredients                   object
n_ingredients                  int64
calories                     float64
total_fat                    float64
sugar                        float64
sodium                       float64
protein                      float64
saturated_fat                float64
carbohydrates                float64
year                           int32
month                          int32
description_filled            object
description_length             int64
dtype: object

In [95]:
from src.data_loader import save_parquet
save_parquet(recipes_enriched, "recipes_enriched.parquet")

💾 data\processed\recipes_enriched.parquet  shape=(231637, 23)


WindowsPath('C:/Users/ABDILLAHIOMARDJAMA/Downloads/EDA_Project/data/processed/recipes_enriched.parquet')